### PR TITLE
refactor: off-chain-auth

### PR DESCRIPTION
**Pull request by ruojunm**
### Description

This pr refactor the off-chain-auth to module.

### Rationale

Refactor

### Example

NA

### Changes

Notable changes: 
* NA

### CODE REVIEW

File: core/module/modular.go [hunk 1]
```diff
@@ -52,6 +54,12 @@ type Authorizer interface {
 	Modular
 	// VerifyAuthorize verifies the operator authority.
 	VerifyAuthorize(ctx context.Context, auth AuthOpType, account, bucket, object string) (bool, error)
+	// GetAuthNonce get the auth nonce for which the Dapp or client can generate EDDSA key pairs.
+	GetAuthNonce(context.Context, *gfspserver.GetAuthNonceRequest) (*spdb.OffChainAuthKey, error)
```
Comment (joeylichang) [inline core/module/modular.go:58]: core dir should use the interface, include the func params
core dir should not import base/types/gfspserver pkg.
suggest to refine gfspserver.GetAuthNonceRequest to interface and gfspserver.GetAuthNonceRequest implement the interface

File: core/module/modular.go [hunk 2]
```diff
@@ -52,6 +54,12 @@ type Authorizer interface {
...
+	// UpdateUserPublicKey updates the user public key once the Dapp or client generates the EDDSA key pairs.
+	UpdateUserPublicKey(context.Context, *gfspserver.UpdateUserPublicKeyRequest) (bool, error)
```
Comment (joeylichang) [inline core/module/modular.go:60]: ditto

File: core/module/modular.go [hunk 3]
```diff
@@ -52,6 +54,12 @@ type Authorizer interface {
...
+	// VerifyOffChainSignature verifies the signature signed by user's EDDSA private key.
+	VerifyOffChainSignature(context.Context, *gfspserver.VerifyOffChainSignatureRequest) (bool, error)
```
Comment (joeylichang) [inline core/module/modular.go:62]: ditto

File: base/gfspapp/authorizer_server.go [hunk 4]
```diff
@@ -27,3 +27,46 @@ func (g *GfSpBaseApp) GfSpVerifyAuthorize(
...
+	return &gfspserver.UpdateUserPublicKeyResponse{
+		Err:    gfsperrors.MakeGfSpError(err),
+		Result: resp,
+	}, nil
+}
+
+// VerifyOffChainSignature verifies the signature signed by user's EDDSA private key.
+func (g *GfSpBaseApp) VerifyOffChainSignature(ctx context.Context, req *gfspserver.VerifyOffChainSignatureRequest) (*gfspserver.VerifyOffChainSignatureResponse, error) {
+
+	log.CtxDebugw(ctx, "begin to VerifyOffChainSignature", "user", req.GetAccountId(), "domain", req.GetDomain(), "OffChainSig", req.OffChainSig, "RealMsgToSign", req.RealMsgToSign)
```
Comment (joeylichang) [inline base/gfspapp/authorizer_server.go:65]: log foemat "OffChainSig" -> "off_chain_sig" , log specs should not use camelCase

File: base/gfspclient/authorizer.go [hunk 5]
```diff
@@ -32,3 +33,52 @@ func (s *GfSpClient) VerifyAuthorize(ctx context.Context,
 	}
 	return resp.GetAllowed(), nil
 }
+
+// GetAuthNonce get the auth nonce for which the Dapp or client can generate EDDSA key pairs.
+func (s *GfSpClient) GetAuthNonce(ctx context.Context, in *gfspserver.GetAuthNonceRequest, opts ...grpc.CallOption) (*gfspserver.GetAuthNonceResponse, error) {
```
Comment (joeylichang) [inline base/gfspclient/authorizer.go:38]: GfSpClient should seem like SDK, the GfSpClient will be offer to other repo, it should not use the gfspserver.GetAuthNonceRequest(Grpc specs), may be gfspapp support other protocol(http/ other rpc), so the client params should use the meaningful params, eg:  accountId, domain.

File: base/gfspclient/authorizer.go [hunk 6]
```diff
@@ -32,3 +33,52 @@ func (s *GfSpClient) VerifyAuthorize(ctx context.Context,
...
+	ctx = log.Context(ctx, resp)
+	if err != nil {
+		log.CtxErrorw(ctx, "failed to get auth nonce rpc", "error", err)
+		return nil, err
+	}
+	return resp, nil
+}
+
+// UpdateUserPublicKey updates the user public key once the Dapp or client generates the EDDSA key pairs.
+func (s *GfSpClient) UpdateUserPublicKey(ctx context.Context, in *gfspserver.UpdateUserPublicKeyRequest, opts ...grpc.CallOption) (*gfspserver.UpdateUserPublicKeyResponse, error) {
```
Comment (joeylichang) [inline base/gfspclient/authorizer.go:55]: ditto

File: base/gfspclient/authorizer.go [hunk 7]
```diff
@@ -32,3 +33,52 @@ func (s *GfSpClient) VerifyAuthorize(ctx context.Context,
...
+	ctx = log.Context(ctx, resp)
+	if err != nil {
+		log.CtxErrorw(ctx, "failed to update user public key rpc", "error", err)
+		return nil, err
+	}
+	return resp, nil
+}
+
+// VerifyOffChainSignature verifies the signature signed by user's EDDSA private key.
+func (s *GfSpClient) VerifyOffChainSignature(ctx context.Context, in *gfspserver.VerifyOffChainSignatureRequest, opts ...grpc.CallOption) (*gfspserver.VerifyOffChainSignatureResponse, error) {
```
Comment (joeylichang) [inline base/gfspclient/authorizer.go:71]: ditto

File: modular/gater/admin_handler.go [hunk 8]
```diff
@@ -46,7 +46,7 @@ func (g *GateModular) getApprovalHandler(w http.ResponseWriter, r *http.Request)
 		log.CtxDebugw(reqCtx.Context(), reqCtx.String())
 	}()
 
-	reqCtx, err = NewRequestContext(r)
+	reqCtx, err = g.NewRequestContext(r)
```
Comment (joeylichang) [inline modular/gater/admin_handler.go:49]: NewRequestContext(r, g), let RequestContext combining resources of g is better?
RequestContext seems not belong to g from function view.

File: base/gfspapp/authorizer_server.go (hunk 4 above)
Comment (ruojunm) [reply base/gfspapp/authorizer_server.go]: Done

File: core/module/modular.go (hunk 2 above)
Comment (ruojunm) [reply core/module/modular.go]: Done

File: base/gfspclient/authorizer.go (hunk 5 above)
Comment (ruojunm) [reply base/gfspclient/authorizer.go]: Done

File: base/gfspclient/authorizer.go (hunk 7 above)
Comment (ruojunm) [reply base/gfspclient/authorizer.go]: Done

File: base/gfspclient/authorizer.go (hunk 6 above)
Comment (ruojunm) [reply base/gfspclient/authorizer.go]: Done

File: core/module/modular.go (hunk 3 above)
Comment (ruojunm) [reply core/module/modular.go]: Done

File: core/module/modular.go (hunk 1 above)
Comment (ruojunm) [reply core/module/modular.go]: Done

File: modular/gater/admin_handler.go (hunk 8 above)
Comment (ruojunm) [reply modular/gater/admin_handler.go]: Done

**Review by joeylichang — APPROVED** [review]

**Review by will-2012 — APPROVED** [review]

**Pull request merged**